#19 feat(Follow) follow feature

uridonget opened the pull request:
# feat(Follow) follow feature \#19
## Overview
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
1. Follow 모댈 추가
2. 닉네임으로 다른 유저를 검색하는 함수 추가
3. 다른 유저를 팔로우 하는 함수 추가
4. 팔로우 하고있는 유저를 언팔로우(팔로우 취소)하는 함수 추가

<!---- Resolves: #(Isuue Number) -->
## PR Type
어떤 변경 사항이 있나요?

- [x] feat

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.

uridonget committed:
#19 Feat(Follow): Follow 모델 추가

1. models.py 파일에 Follow 클래스 추가.
2. admin.py 파일에 FollowAdmin 클래스 추가.

uridonget committed:
#19 Feat(Follow): 유저를 닉네임으로 탐색하는 함수 추가

search_user_by_nickname 함수를 추가하였음.

uridonget committed:
#19 Chore(Follow): 유저를 닉네임으로 탐색하는 함수 수정

1. search_user_by_nickname 함수를 수정하였음.
    * 기능 수정사항
    - 수정 전 : 특정 닉네임과 일치하는 닉네임을 가진 유저가 존재하는지 탐색.
    - 수정 후 : 특정 단어가 포함된 닉네임을 가진 유저가 존재하는지 탐색.

    * 코드 수정
    - 클래스형 뷰로 수정.

2. search_user_by_nickname 함수를 호출하는 urls endpoint 추가

uridonget committed:
#19 Fix(JWT): JWT 인증함수(authenticate_token) 수정

access_token이 소실되는 경우에 대한 처리를 추가하였음.

uridonget committed:
#19 Feat(CSRF): CSRF 토큰을 발급하는 함수 추가

- 프론트 서버에서 CSRF 토큰을 요청했을 때 토큰을 발급하는 get_csrf_token 함수를 추가하였음.
- CSRF 관련 설정들을 settings.py 파일에 추가하였음.

uridonget committed:
#19 Feat(CSRF): CSRF 토큰을 발급하는 함수 추가

- 프론트 서버에서 CSRF 토큰을 요청했을 때 토큰을 발급하는 get_csrf_token 함수를 추가하였음.
- CSRF 관련 설정들을 settings.py 파일에 추가하였음.

uridonget committed:
#19 Feat(Follow): 팔로우 함수 추가

- 특정 유저를 팔로우 하는 함수를 추가하였음. (Follow 클래스의 post method)
- @csrf_protect 데코레이터를 사용하여 csrf 검증을 추가하였음.

uridonget committed:
#19 Feat(Follow): 팔로우 목록을 불러오는 함수 추가

- accout/views.py 파일에 get_follows 함수를 추가하였음.
- 유저가 팔로우하고있는 모든 유저의 데이터를 JsonResponse 형태로 반환함.

uridonget committed:
#19 Feat(Follow): 팔로우를 취소하는 함수 추가

- 특정 유저를 언팔로우 하는 함수를 추가하였음. (Follow 클래스의 delete method)
- 팔로우 정보를 가지고 있는 Follow 클래스의 특정 객체를 삭제함.

uridonget committed:
#19 Fix(Follow): Follow 모델 수정

* 문제 상황
- Follow 모델에서 userAId와 userBId가 OneToOneField로 설정되어 있어
  하나의 유저가 단 한 명의 유저만 팔로우 할 수 있도록 제한되어 있음.

* 수정
- userAId와 userBId를 ForeignKey로 변경하여 한 유저가 여러 명을 팔로우 할 수 있도록 수정하였음.
- ForeignKey로 User 객체를 직접 가르키기에 userAId를 userA로 수정하였음.(마찬가지로 userBId -> userB)

uridonget committed:
#19 Chore(Follow): get_follows 함수 수정

- Follow 테이블을 순회해서 특정 유저의 팔로워들을 찾는 기존의 방식을 수정하였음
- Follow 테이블의 userA related name을 통해서 팔로워를 한 번에 불러올 수 있음.
- 마찬가지의 방법으로 나를 팔로우 하는 사람들의 목록도 불러올 수 있으나 기획상에서 필요가 없기에 추가하지 않았음.

uridonget committed:
merge-conflict fixed

uridonget committed:
merge-conflict fixed(2)

allorak333 commented on backend/account/views.py:
```diff
@@ -539,3 +550,179 @@ def logout(request):
     return response
+
+
+@ensure_csrf_cookie
+def get_csrf_token(request):
+    """
+    @brief 요청할 때마다 새로운 CSRF 토큰을 반환하는 API
+
+    @param request Django의 HTTP 요청 객체
+
+    @return CSRF 토큰을 포함한 JSON 응답
+    """
+    return JsonResponse({"csrfToken": get_token(request)})
```
저희 로그인 로직에는 jwt를 cookie에 저장(http only)하는 방식을 사용하고 있기에 session을 않는 곳에 csrf_token 을 적용할 필요는 없어 보입니다. 2FA와 email 인증에는 session을 사용하기에 이 부분에 대해서만 사용해야하는지 다시 정해보면 될 것 같습니다!

allorak333 commented on backend/account/views.py:
```diff
@@ -539,3 +550,179 @@ def logout(request):
+            if not word:
+                return JsonResponse({"message": "Word not provided"}, status=400)
+
+            followings = User.objects.filter(nickname__icontains=word)
```
아직 follow 되지 않은 유저 목록이기에 follwoings보다는 user_list라 하는 것이 더 좋을 것 같습니다!

allorak333 commented on backend/transcendence/settings.py:
```diff
@@ -172,7 +172,12 @@
 
 # CORS
-CORS_ALLOW_ALL_ORIGINS = True
+CORS_ALLOW_ALL_ORIGINS = [
+    os.environ.get("BACK_SERVER_URL"),
+    os.environ.get("FRONT_SERVER_URL"),
+]
```
CORS_ALLOW_ALL_ORIGIN 보다는 CORS_ALLOWED_ORIGINS 으로 바꾸는 것이 의도에 맞을 것 같습니다.

uridonget committed:
#19 Chore(Follow): PR feedback 수정

- 세션을 사용하지 않는 함수에서의 csrf 검사를 생략하였음.
- follow의 get 요청처리 함수에서 적절하지 않은 변수명을 수정하였음.
- settings.py에서 잘못된 설정을 수정하였음.

allorak333 reviewed (approved):
LGTM!
